### ISSUE TITLE
BUG,ENH: cython stubs should use `except (0|-1?)` etc. for integer returning functions

### ISSUE THREAD

**Issue by zoj613** [numpy/numpy#19291]
According to the docs `PyArray_IntpConverter` is supposed to just return `NPY_FAIL` (0) on failure and `NPY_SUCCESS` (1) otherwise. So to help manually raise an exception I write something like:
```cython
cdef np.PyArray_Dims shape
if not np.PyArray_IntpConverter(size, &shape):
	raise ValueError("`size` is not a valid argument.")
```

But this ends up raising a `SystemError` with the message: `<class 'ValueError'> returned a result with an error set`, Which ends up causing unittests to fail since the intended ValueError is not raised.
### Reproducing code example:
Passing a set for `size` instead of the expected tuple.

Is there a way to silence the `SystemError` and only raise the intended one?

### NumPy/Python version information:

```
1.19.0 3.6.12 (default, May 11 2021, 17:48:34) 
[GCC 10.2.0]
```

**Comment by seberg** [comment]
An exception must be/is set if a converter fails.  It may be that the cython `.pyd` file is wrong though, and might be nicer to read: `int PyArray_IntpConverter(...) except 0`.  And then be used with a try/except branch if you want to replace the error.

Although, I guess there may be ways to explicitly probe an in-fly error here. (If just by calling `PyErr_Occurred()` explicitly)

**Comment by zoj613** [comment]
The declaration in `__init__.pxd` currently is
```cython
int PyArray_IntpConverter (object, PyArray_Dims *)
```

Looking at the function implementation, it seems that the exception originates from this line:
https://github.com/numpy/numpy/blob/20e14f3a6a2368d240820a4b6293f0837406c087/numpy/core/src/multiarray/conversion_utils.c#L142
which maybe the one responsible for the "silent" exception. I supposed the `PyErr_Occurred()` would be added in the following block, right?

Also just noticed this is related to: https://github.com/numpy/numpy/issues/19010, I guess I no longer have a reason to procrastinate now :D

**Comment by seberg** [comment]
No, it is unrelated to gh-19010, because the behaviour is absolutely correct.  The `__init__.pxd` may be wrong, but from a C-perspective, it is exactly as advertised.

**Comment by zoj613** [comment]
> No, it is unrelated to [gh-19010](https://github.com/numpy/numpy/issues/19010), because the behaviour is absolutely correct. The `__init__.pxd` may be wrong, but from a C-perspective, it is exactly as advertised.

Not the best wording there from my last reply. I meant that the function is directly related to the ones that need to be addressed in https://github.com/numpy/numpy/issues/19010, not the issue itself. What do you suggest is the correct course of action to address this one?

**Comment by zoj613** [comment]
@seberg By the way, [PyDimMem_FREE](https://numpy.org/devdocs/reference/c-api/array.html#c.PyDimMem_FREE) is not available as part of the cython declarations and is required by [PyArray_IntpConverter](https://numpy.org/devdocs/reference/c-api/array.html#c.PyArray_IntpConverter) according to the docs, to free the pointer in `shape`. I use `PyMem_Free` as a workaround currently but maybe this also should be addressed? Or at least the docs be updated?

**Comment by seberg** [comment]
I am pretty sure that `PyMem_Free` is incorrect (a debug python will shout at you, a non-debug python might just crash randomly at some point). `PyMem_RawFree` or just `free` is probably right.  Ideally we would force you to use `PyDimMem_FREE` (and make it a function!), so that we could use `PyMem_Free` or similar at some point.  The fact that this function allocates the dimensions at all is a bit annoying IMO.

Honestly, I am not sure what to do best here.  For all I know the cython definitions should be adapted to way I said.  At least unless the current version is actually useful...  `PyDimMem_FREE` should maybe be exposed (I don't particularly like the API, there is no reason the caller shouldn't just stack allocate that array to begin with, it would suit most usages better).

**Comment by zoj613** [comment]
Thanks for the suggestion, I wasn't quite sure which free function to pick. That helps a lot. Maybe a new function that allows the caller to allocate the memory and pass a pointer into it is a possibility? Something like:
```
int PyArray_IntpConverter2(PyObject *obj, npy_intp* dim_ptr, PyArray_Dims *seq)
```
which will do the length calculation and then assign the `ptr` member of `seq` struct to the `dim_ptr` buffer.

**Comment by mattip** [comment]
What is the bigger picture: why are you parsing shapes? Perhaps the problem is that we expose `PyArray_IntpConverter` and `PyArray_Dims` at all.

**Comment by zoj613** [comment]
> What is the bigger picture: why are you parsing shapes? Perhaps the problem is that we expose `PyArray_IntpConverter` and `PyArray_Dims` at all.

To return a shape broadcasted array, as in functions like `normal([1, 2, 3], size=(2, 4, 3)) -> (2, 4, 3). To do this using the C-API does not seem very easy. One way I did this is to:
`create an empty (2, 4, 3) array -> broadcast that with the input using PyArray_MultiIterNew3() to get a new broadcast object and then using that to populate the return array.` I dont like this way because it wastes memory. The other is to use `PyArray_IntpConverter` on the input shape to create a `shape` variable of type `PyArray_Dims` that can be used with `PyArray_BroadcastToShape` on the individual inputs to create iterators. This way I dont have to create an intermediate empty array just so I can broadcast using the multi-iterator. Maybe there are functions I am not aware of but broadcasting using the C-API seems a bit involved hence why I require this approach. If you know a better way to do this using the C-API i'd be happy to hear it.

**Comment by mattip** [comment]
> functions like normal([1, 2, 3], size=(2, 4, 3)) -> (2, 4, 3)

Can you link to this function? I still don't get what the goal is. If you want to return a shape-broadcast array via a call to a numpy function, just call the function and it will return the array? I am missing something basic.

**Comment by zoj613** [comment]
> > functions like normal([1, 2, 3], size=(2, 4, 3)) -> (2, 4, 3)
> 
> Can you link to this function? I still don't get what the goal is. If you want to return a shape-broadcast array via a call to a numpy function, just call the function and it will return the array? I am missing something basic.

Sure, Here is the function: https://github.com/zoj613/polyagamma/blob/86585f629c5260f4c6d7b6da5a996d064dfaa28b/polyagamma/_polyagamma.pyx#L114-L323

**Comment by zoj613** [comment]
Note that, I would have much rather used the functions provided in numpy.random._common.pxd but that file is not public. So I had to implement the logic myself and prefer sticking to the C-API instead of calling the high level python functions which introduce unwanted overhead.

**Comment by mattip** [comment]
Well, you can still copy the code from the way `_common.pyx` [handles this](https://github.com/numpy/numpy/blob/main/numpy/random/_common.pyx#L492).

**Comment by mattip** [comment]
.. or with one input, I think [this is the right function](https://github.com/numpy/numpy/blob/main/numpy/random/_common.pyx#L418)

**Comment by bashtage** [comment]
> So I had to implement the logic myself and prefer sticking to the C-API instead of calling the high level python functions which introduce unwanted overhead.

Have you benchmarked the overhead for the limited use of the Python API here?  Except in the smallest broadcasting cases I suspect it is undetectable.

**Comment by bashtage** [comment]
@zoj613 Where do you get the idea that there are two allocations in that function?  There is only one array allocation for each of the possible branches. If the user passes a size, then this size is taken as fact for he output size (but is checked later for correctness).  If the user doesn't pass a size, then an iterator is created to determine the size of the array.

The only place where I know the pattern you are using is taken is in some of the more complex multidimensional broadcasting code. See https://github.com/numpy/numpy/pull/16740/files#diff-47f9acf67023d339efc93e1a1123939a182e6d2458c22fcf783cd98d75b71675R3846 for example where a temp array is created to then get the broadcast size.

**Comment by zoj613** [comment]
> then this size is taken as fact for he output size (but is checked later for correctness).

Yes, you are right. There is one allocation. My issue is that this approach encourages use of the high-level python functions to get the desired results. My main motivator for using `PyArray_InpConverter` is so that I can have the information about dimensions and shape in a C-workable way. I could write things like this:
```cython
    if size:
        if PyTuple_CheckExact(size):
            shape = np.PyArray_FROM_OT(size, np.NPY_INTP)
            empty_arr = np.PyArray_EMPTY(
                PyTuple_GET_SIZE(size),
                <np.npy_intp*>np.PyArray_DATA(shape),
                np.NPY_DOUBLE,
                0
            )
        else:
            arr_len = <np.npy_intp>size
            empty_arr = np.PyArray_EMPTY(1, &arr_len, np.NPY_DOUBLE, 0)
        bcast = np.PyArray_MultiIterNew3(h, z, empty_arr)
        arr_ptr = <double*>np.PyArray_DATA(empty_arr)

        with bitgenerator.lock, nogil:
            while bcast.index < bcast.size:
                ch = (<double*>np.PyArray_MultiIter_DATA(bcast, 0))[0]
                cz = (<double*>np.PyArray_MultiIter_DATA(bcast, 1))[0]
                arr_ptr[bcast.index] = whatever_func(bitgen, ch, cz, stype)
                np.PyArray_MultiIter_NEXT(bcast)
        return arr
```
But this is overly verbose and could be avoided by making use of `PyArray_IntpConverter` like so:

```cython
cdef np.PyArray_Dims shape
if not np.PyArray_IntpConverter(size, &shape):
	raise ValueError("`size` is not a valid argument.")
else:
    empty_arr = np.PyArray_EMPTY(shape.len, shape.ptr, np.NPY_DOUBLE, 0)
    bcast = np.PyArray_MultiIterNew3(h, z, empty_arr)
    arr_ptr = <double*>np.PyArray_DATA(empty_arr)

    with bitgenerator.lock, nogil:
        while bcast.index < bcast.size:
          	ch = (<double*>np.PyArray_MultiIter_DATA(bcast, 0))[0]
            cz = (<double*>np.PyArray_MultiIter_DATA(bcast, 1))[0]
            arr_ptr[bcast.index] = whatever_func(bitgen, ch, cz, stype)
            np.PyArray_MultiIter_NEXT(bcast)
    return arr
```

The issue is that when `size` is not the appropriate type (e.g someone passing in a set), the exception is not raised properly. Instead of a ValueError I get a `SystemError`, which suggests that an exception was raised silently before getting to the `ValueError` line.

**Comment by bashtage** [comment]
> My issue is that this approach encourages use of the high-level python functions to get the desired results

I would really encourage you to benchmark these rather than simply assuming that a single (or even a few) are particularly expensive.  IME it is a bit of a myth that single calls to Python functions inside a python-accessible function are very slow.

A simple workaround for your shorter code would be to call `PyErr_Clear` before raising ValueError.

**Comment by eric-wieser** [comment]
How does cython handle `Converter` functions in the the C API? The only one I can find that looks public is `PyUnicode_FSConverter`, and cython seems not to mention it anyway.

**Comment by bashtage** [comment]
FWIW, from reading the document, if `np.PyArray_IntpConverter(size, &shape)` succeeds then you need to call free on `shape->ptr` to avoid a leak. after you initialize your array.

**Comment by zoj613** [comment]
> > My issue is that this approach encourages use of the high-level python functions to get the desired results
> 
> I would really encourage you to benchmark these rather than simply assuming that a single (or even a few) are particularly expensive. IME it is a bit of a myth that single calls to Python functions inside a python-accessible function are very slow.
> 
> A simple workaround for your shorter code would be to call `PyErr_Clear` before raising ValueError.

I did benchmark them locally quite a few times. This is how I ended up with the current implementation. The others are slower. I will try the `PyErr_Clear` approach.

> FWIW, from reading the document, if `np.PyArray_IntpConverter(size, &shape)` succeeds then you need to call free on `shape->ptr` to avoid a leak. after you initialize your array.
I do call free in the actual code after I create the empty array.

**Comment by bashtage** [comment]
> I did benchmark them locally quite a few times. This is how I ended up with the current implementation. The others are slower. I will try the `PyErr_Clear` approach.

Noticeably slower or just slower?

**Comment by zoj613** [comment]
> > I did benchmark them locally quite a few times. This is how I ended up with the current implementation. The others are slower. I will try the `PyErr_Clear` approach.
> 
> Noticeably slower or just slower?

Well, this one is subjective. To me it is slower because if I can make it faster while writing it cleaner then it is worth it. Slower to me is `7 > microseconds` difference between the two snippets above when called with the argument: `(h=[2, 3, 4], size=(10, 5, 3))`. It gets more noticeable with large arrays and `size` elements.

**Comment by bashtage** [comment]
To me the actual iteration across the broadcast is the most expensive part. For example,

```
In [28]: %timeit g.normal([2,3,4],size=(10,10,10,5,3))
292 µs ± 1.22 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [29]: %timeit g.normal(0,size=(10,10,10,5,3))
171 µs ± 1.02 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

The first path uses the broadcast iterator while the second does not.

**Comment by zoj613** [comment]
> To me the actual iteration across the broadcast is the most expensive part. For example,
> 
> ```
> In [28]: %timeit g.normal([2,3,4],size=(10,10,10,5,3))
> 292 µs ± 1.22 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
> 
> In [29]: %timeit g.normal(0,size=(10,10,10,5,3))
> 171 µs ± 1.02 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
> ```
> 
> The first path uses the broadcast iterator while the second does not.

You're comparing the same implementation for different input. It would be interesting to see what the results are if you were to rerun it using `PyArray_InpConverter` without the intermediate python calls.

**Comment by mattip** [comment]
I think "correct" is more important than "fast". The `PyArray_InpConverter` path is apparently is very difficult to do correctly.

**Comment by zoj613** [comment]
> I think "correct" is more important than "fast". The `PyArray_InpConverter` path is apparently is very difficult to do correctly.

Isn't this a simple fix of making sure PyErr_Occured() is called if `PyArray_IntpFromIndexSequence` fails inside of `PyArray_IntpFromIndexSequence`? Outside of that, the function works as expected as long as the input can pass through `PyArray_IntpFromIndexSequence`. Or am I misinterpreting things?

**Comment by mattip** [comment]
This comment about freeing memory is true in that `PyMem_Free` is incorrect.

> I am pretty sure that `PyMem_Free` is incorrect (a debug python will shout at you, a non-debug python might just crash randomly at some point). `PyMem_RawFree` or just free is probably right.

NumPy tries to pool allocations. There is a specific pool for shape/strides. The function calls `npy_alloc_cache_dim` underneath to allocate from the pool, and as far as I can tell there is no way to call  `npy_free_cache_dim` so using it via the C-API breaks the cache.

**Comment by zoj613** [comment]
> This comment about freeing memory is true in that `PyMem_Free` is incorrect.
> 
> > I am pretty sure that `PyMem_Free` is incorrect (a debug python will shout at you, a non-debug python might just crash randomly at some point). `PyMem_RawFree` or just free is probably right.
> 
> NumPy tries to pool allocations. There is a specific pool for shape/strides. The function calls `npy_alloc_cache_dim` underneath to allocate from the pool, and as far as I can tell there is no way to call `npy_free_cache_dim` so using it via the C-API breaks the cache.

Are you suggesting that the documentation is not correct about requiring `PyDimMem_FREE` to free the memory block? As far as I can tell, it seems the macros is warpping libc's `free` https://github.com/numpy/numpy/blob/fc8e3bbe419748ac5c6b7f3d0845e4bafa74644b/numpy/core/include/numpy/ndarraytypes.h#L383

**Comment by seberg** [comment]
Honestly, aside of the reasonable question whether using these converters is actually *useful* here...  I think the main question should just be whether the current definition is not just an outright bug.

The `.pxd` should read:
```
cdef int ConverterFunction(...) except 0
```
and after that change, I am sure you can figure out the correct cython code.  Maybe just do that manually.

Some of our converters are there for a reason, and different from the default converters.  So it might be nice to use them, although they are really not designed to be called directly, normally.  But then they are designed for the Python C-API.

>  as far as I can tell there is no way to call npy_free_cache_dim so using it via the C-API breaks the cache.

Probably, but it is actually completely fine to *not* use `npy_free_cache_dim`. `PyDimMem_FREE` is valid, the cache API is designed to support it.  And I think its just a macro around `PyArray_free`, which is just a macro for `free`/`PyMem_RawFree`, which are functionally equivalent probably.

**Comment by seberg** [comment]
With a good chance I am pinging the wrong person, but @scoder what is the right thing here?

NumPy has a couple of functions with integer returns that are only used to indicate error resutls, examples are:
```
int PyArray_CopyObject (ndarray, object)
int PyArray_Sort (ndarray, int, NPY_SORTKIND)
int PyArray_IntpConverter (object, PyArray_Dims *)  # this issue
```
which only have a return value to indicate an error result (normally `except -1`, converters with `except 0`).  Now, I don't condone half of those functions (they just don't look like they should be public).  But my question is if the missing `except -1` could just be changed as a bug-fix or whether it is more complicated?

**Comment by scoder** [comment]
> my question is if the missing `except -1` could just be changed as a bug-fix or whether it is more complicated?

It can just be added and will make the user code safer in the future. Code that manually handles errors now will become unused, which shouldn't hurt because it was most likely fragile at best previously. These functions are very unlikely to be the only source of exceptions in any given piece of Cython code.

Note that it needs to be added in NumPy first, then ported over to Cython (which now only keeps a legacy copy of the NumPy declarations).

We can also fix this in the 0.29.x release series.

**Comment by zoj613** [comment]
@seberg Has there been a decision reached regarding the list of functions you mention in your last comment?
